msteams-platform/samples/virtual-assistant.md

```diff
@@ -0,0 +1,560 @@
+
+![Confirmation prompt when switching to a new skill](../assets/images/bots/virtual-assistant/switch-skills-prompt.png)
+
+### Handling task module requests
+
+To add task module capabilities to a Virtual Assistant, two additional methods are included in the Virtual Assistant activity handler: `OnTeamsTaskModuleFetchAsync` and `OnTeamsTaskModuleSubmitAsync`. These methods listen to task module-related activities from Virtual Assistant, identify the skill associated with the request, and forward the request to the identified skill. 
+
+Request forwarding is done via the  [SkillHttpClient](/dotnet/api/microsoft.bot.builder.integration.aspnet.core.skills.skillhttpclient?view=botbuilder-dotnet-stable), `PostActivityAsync` method. It returns the response as `InvokeResponse` which is parsed and converted to `TaskModuleResponse` .
+
+```csharp
+    public static TaskModuleResponse GetTaskModuleRespose(this InvokeResponse invokeResponse)
+    {
+        if (invokeResponse.Body != null)
+        {
+            return new TaskModuleResponse()
+            {
+                Task = GetTask(invokeResponse.Body),
+            };
+        }
+
+        return null;
+    }
+
+    private static TaskModuleResponseBase GetTask(object invokeResponseBody)
+        {
+            JObject resposeBody = (JObject)JToken.FromObject(invokeResponseBody);
+            var task = resposeBody.GetValue("task");
+            var taskType = task.SelectToken("type").ToString();
+
+            return taskType switch
+            {
+                "continue" => new TaskModuleContinueResponse()
+                {
+                    Type = taskType,
+                    Value = task.SelectToken("value").ToObject<TaskModuleTaskInfo>(),
+                },
+                "message" => new TaskModuleMessageResponse()
+                {
+                    Type = taskType,
+                    Value = task.SelectToken("value").ToString(),
+                },
+                _ => null,
+            };
+        }
+```
+
+A similar approach is followed for card action dispatching and task module responses. Task module fetch and submit action data is updated to include `skillId`. 
+Activity Extension method `GetSkillId` extracts `skillId` from the payload which provides details about the skill that needs to be invoked.
+
+Below is a code snippet for `OnTeamsTaskModuleFetchAsync` and `OnTeamsTaskModuleSubmitAsync` methods.
+
+```csharp
+    // Invoked when a "task/fetch" event is received to invoke task module.
+    protected override async Task<TaskModuleResponse> OnTeamsTaskModuleFetchAsync(ITurnContext<IInvokeActivity> turnContext, TaskModuleRequest taskModuleRequest, CancellationToken cancellationToken)
+    {
+        try
+        {
+            string skillId = (turnContext.Activity as Activity).GetSkillId();
+            var skill = _skillsConfig.Skills.Where(s => s.Value.AppId == skillId).FirstOrDefault().Value;
```

Comment:
FirstOrDefault() can return null if there's no matching skill, which will then throw a NullReferenceException when we try to get `.Value`. If the exception here is expected, we might as well use First().

Reply:
Same comment for the submit method below

Reply:
Exception is catched and proper error message is shown. Corrected the `FirstOrDefault()` to `First()`.